src/plugins/follow_config/FollowConfig.cc

```diff
@@ -0,0 +1,265 @@
+         << this->dataPtr->followTargetName << "]" << std::endl;
+      this->dataPtr->newFollowUpdateTargetName = true;
+    }
+    if (auto offsetElem = _pluginElem->FirstChildElement("follow_offset"))
+    {
+      std::stringstream offsetStr;
+      offsetStr << std::string(offsetElem->GetText());
+      offsetStr >> this->dataPtr->followOffset;
+      gzmsg << "FollowConfig: Loaded follow_offset from sdf ["
+         << this->dataPtr->followOffset << "]" << std::endl;
+      this->dataPtr->newFollowUpdateOffset = true;
+    }
+    if (auto pGainElem = _pluginElem->FirstChildElement("follow_pgain"))
+    {
+      this->dataPtr->followPGain = std::stod(std::string(pGainElem->GetText()));
+      gzmsg << "FollowConfig: Loaded follow_pgain from sdf ["
+         << this->dataPtr->followPGain << "]" << std::endl;
+      this->dataPtr->newFollowUpdatePGain = true;
+    }
+  }
+
+  gui::App()->findChild<
+      MainWindow *>()->installEventFilter(this);
+}
+
+/////////////////////////////////////////////////
+bool FollowConfig::eventFilter(QObject *_obj, QEvent *_event)
+{
+  if (_event->type() == events::Render::kType)
+  {
+    if (this->dataPtr->newFollowUpdateTargetName)
+    {
+      this->dataPtr->UpdateFollowTargetName();
+    }
+    if (this->dataPtr->newFollowUpdatePGain)
+    {
+      this->dataPtr->UpdateFollowPGain();
+    }
+    if (this->dataPtr->newFollowUpdateOffset)
+    {
+      this->dataPtr->UpdateFollowOffset();
+    }
+  }
+
+  // Standard event processing
+  return QObject::eventFilter(_obj, _event);
+}
+
+/////////////////////////////////////////////////
+void FollowConfig::SetFollowOffset(double _x,
+          double _y, double _z)
+{
+  if (!this->dataPtr->newFollowUpdateOffset)
+  {
+    this->dataPtr->followOffset = math::Vector3d(
+      _x, _y, _z);
+      gzmsg << "FollowConfig: SetFollowOffset("
+        << this->dataPtr->followOffset << ")" << std::endl;
+    this->dataPtr->newFollowUpdateOffset = true;
+  }
+}
+
+/////////////////////////////////////////////////
+void FollowConfig::SetFollowPGain(double _p)
+{
+  if (!this->dataPtr->newFollowUpdatePGain)
+  {
+    this->dataPtr->followPGain = _p;
+    gzmsg << "FollowConfig: SetFollowPGain("
+         << this->dataPtr->followPGain << ")" << std::endl;
+    this->dataPtr->newFollowUpdatePGain = true;
+  }
+}
+
+/////////////////////////////////////////////////
+void FollowConfigPrivate::UpdateFollowTargetName()
+{
+  // Offset
+  std::function<void(const msgs::Boolean &, const bool)> cbName =
+    [&](const msgs::Boolean &/*_rep*/, const bool _resultName)
+  {
+    if (!_resultName)
+    {
+      gzerr << "FollowConfig: Error sending follow target name." << std::endl;
+    } else {
+      gzmsg << "FollowConfig: Request Target Name: "
+            << this->followTargetName << " sent" << std::endl;
+    }
+  };
+
+  msgs::StringMsg reqName;
+  reqName.set_data(this->followTargetName);
+  node.Request(this->followTargetNameService, reqName, cbName);
```

Comment:
If you're using async service requests, you'll need to synchronize access to all the variables used by the callback.

Reply:
Alternatively, just copy-capture the variable in the lambda. I think this is a lot easier.

Reply:
Will finally get back to looking at this.

Reply:
@bperseghetti did you address the feedback ?